src/test/java/com/metamx/common/parsers/JSONPathParserTest.java

```diff
@@ -0,0 +1,213 @@
+    Assert.assertNull(jsonMap.get("ignore_me"));
+    Assert.assertNull(jsonMap.get("foo"));
+
+    // Invalid fields
+    Assert.assertNull(jsonMap.get("INVALID_ROOT"));
+    Assert.assertNull(jsonMap.get("INVALID_PATH"));
+  }
+
+  @Test
+  public void testNestingNoDiscovery()
+  {
+    List<JSONPathParser.FieldSpec> fields = new ArrayList<>();
+    fields.add(new JSONPathParser.FieldSpec(JSONPathParser.FieldType.ROOT, "simpleVal", "simpleVal"));
+    fields.add(new JSONPathParser.FieldSpec(JSONPathParser.FieldType.ROOT, "timestamp", "timestamp"));
+    fields.add(new JSONPathParser.FieldSpec(JSONPathParser.FieldType.PATH, "nested-foo.bar2", "$.foo.bar2"));
+    fields.add(new JSONPathParser.FieldSpec(JSONPathParser.FieldType.PATH, "heybarx0", "$.hey[0].barx"));
+    fields.add(new JSONPathParser.FieldSpec(JSONPathParser.FieldType.PATH, "met-array", "$.met.a"));
+
+    final Parser<String, Object> jsonParser = new JSONPathParser(fields, false, null);
+    final Map<String, Object> jsonMap = jsonParser.parse(nestedJson);
+
+    // Root fields
+    Assert.assertEquals("text", jsonMap.get("simpleVal"));
+    Assert.assertEquals("2999", jsonMap.get("timestamp"));
+
+    // Nested fields
+    Assert.assertEquals("bbb", jsonMap.get("nested-foo.bar2"));
+    Assert.assertEquals("asdf", jsonMap.get("heybarx0"));
+    Assert.assertEquals(ImmutableList.of(7L, 8L, 9L), jsonMap.get("met-array"));
+
+    // Fields that should not be discovered
+    Assert.assertNull(jsonMap.get("newmet"));
+    Assert.assertNull(jsonMap.get("foo.bar1"));
+    Assert.assertNull(jsonMap.get("baz"));
+    Assert.assertNull(jsonMap.get("blah"));
+    Assert.assertNull(jsonMap.get("nested-foo.bar1"));
+    Assert.assertNull(jsonMap.get("hey"));
+    Assert.assertNull(jsonMap.get("met"));
+    Assert.assertNull(jsonMap.get("ignore_me"));
+    Assert.assertNull(jsonMap.get("foo"));
+  }
+
+  @Test
+  public void testRejectDuplicates()
+  {
+    List<JSONPathParser.FieldSpec> fields = new ArrayList<>();
+    fields.add(new JSONPathParser.FieldSpec(JSONPathParser.FieldType.PATH, "met-array", "$.met.a"));
+    fields.add(new JSONPathParser.FieldSpec(JSONPathParser.FieldType.PATH, "met-array", "$.met.a"));
+
+    try {
+      final Parser<String, Object> jsonParser = new JSONPathParser(fields, false, null);
+      final Map<String, Object> jsonMap = jsonParser.parse(nestedJson);
+    } catch (Exception e) {
+      Assert.assertEquals(e.getMessage(), "Cannot have duplicate field definition: met-array");
+      return;
+    }
+    Assert.fail("Duplicate fields did not trigger exception.");
+  }
+
+  @Test
+  public void testParseFail()
+  {
+    List<JSONPathParser.FieldSpec> fields = new ArrayList<>();
+
+    try {
+      final Parser<String, Object> jsonParser = new JSONPathParser(fields, true, null);
+      final Map<String, Object> jsonMap = jsonParser.parse(notJson);
+    } catch (Exception e) {
+      Assert.assertEquals(ParseException.class, e.getClass());
```

Comment:
(non blocking) you can simply specify `@Test(expected = ParseException.class)` or something like that, I forget the exact name.